[ci] Upload test stats to rockset

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74406

This PR uploads test stats to Rockset. The flow is as follows:
1. Each test writes out XML test reports with xmlrunner.
2. We gather those up and upload them to S3 or GHA.
3. When the CI workflow finishes, another workflow will download the test cases from S3/GHA, parse them, and upload them to Rockset.

suo committed:
initial test stats work

[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

janeyx99 commented on tools/stats/upload_test_stats.py:
```diff
@@ -0,0 +1,175 @@
+    return ret
+
+
+def get_artifact_urls(workflow_run_id: int) -> Dict[Path, str]:
```
Ah, is this written because artifact transfer is not a thing for workflow_runs? That would make sense.

janeyx99 commented on tools/stats/upload_test_stats.py:
```diff
@@ -0,0 +1,175 @@
+        test_cases.extend(parse_xml_report(xml_report, int(args.workflow_run_id)))
+
+    # Write the JSON to rockset
+    print(f"Writing {len(test_cases)} test reports to Rockset")
```
```suggestion
    print(f"Writing {len(test_cases)} test cases to Rockset")
```
nit for clarity

janeyx99 commented on tools/stats/upload_test_stats.py:
```diff
@@ -0,0 +1,175 @@
+    client = rockset.Client(
+        api_server="api.rs2.usw2.rockset.com", api_key=os.environ["ROCKSET_API_KEY"]
+    )
+    client.Collection.retrieve("test_run").add_docs(test_cases)
```
🎉

janeyx99 commented on .github/scripts/get_workflow_job_id.py:
```diff
@@ -37,30 +37,16 @@
+)
 
+jobs = response.json()["jobs"]
+while "next" in response.links.keys():
```
I feel like we do this GH querying logic so much we should have a standardized way of doing it (no need to address in this PR though)

janeyx99 commented:
Seems legit so far, looking forward to seeing those stats in Rockset as a test plan :)

suo commented:
They're actually up already! https://console.rockset.com/collections/details/commons.test_run?tab=summary

One thing that I'm kind of annoyed by is that the timestamp isn't timezone-aware. I wrote https://github.com/xmlrunner/unittest-xml-reporting/pull/269 to fix this upstream, but I don't think it's a blocker

janeyx99 commented:
This isn't a comment against your PR (which just parses XML) but it seems that XFails are considered as skips, which intrigued me:
<img width="1284" alt="image" src="https://user-images.githubusercontent.com/31798555/159527520-d2cf4828-ff38-43f7-b46d-01bcb672df47.png">

janeyx99 commented on .github/workflows/_android-build-test.yml:
```diff
@@ -25,22 +25,22 @@ jobs:
       # [see note: pytorch repo ref]
       - name: Checkout PyTorch
-        uses: pytorch/pytorch/.github/actions/checkout-pytorch@master
+        uses: pytorch/pytorch/.github/actions/checkout-pytorch@master@${{ github.sha }}
```
these changes are separated out now, right?

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "initial test stats work"




[ghstack-poisoned]

suo committed:
Update on "[ci] Upload test stats to rockset"


This PR uploads test stats to Rockset. The flow is as follows:
1. Each test writes out XML test reports with xmlrunner.
2. We gather those up and upload them to S3 or GHA.
3. When the CI workflow finishes, another workflow will download the test cases from S3/GHA, parse them, and upload them to Rockset.

[ghstack-poisoned]

atalman commented on tools/stats/upload_test_stats.py:
```diff
@@ -0,0 +1,181 @@
+    # Write the JSON to rockset
+    print(f"Writing {len(test_cases)} test cases to Rockset")
+    client = rockset.Client(
+        api_server="api.rs2.usw2.rockset.com", api_key=os.environ["ROCKSET_API_KEY"]
```
should we move this constant to "api.rs2.usw2.rockset.com" to beginning of the file with other constants ?

janeyx99 reviewed (approved):
Looks p awesome to me!

suo commented:
@pytorchbot merge this
